clang-tidy: added the ability to shard jobs

gaaclarke opened the pull request:
This will allow bots to split up clang tidy work across the 4 instances of clang tidy so shared classes like everything in fml don't get linted 4 times.

This is accomplished by adding the flags: `--shard-id 0 --shard-variants="foo,bar"`.

issue https://github.com/flutter/flutter/issues/114632

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

jmagman commented on tools/clang_tidy/lib/clang_tidy.dart:
```diff
@@ -112,9 +123,16 @@ class ClangTidy {
       options.buildCommandsPath.readAsStringSync(),
     ) as List<dynamic>;
-    final List<Command> changedFileBuildCommands = await getLintCommandsForChangedFiles(
+    final List<List<dynamic>> shardBuildCommandsData = options
```
I know the rest of this file does this, but nit don't use `dynamic` in null safe code, use `Object?` or `Object` if you know it won't be null.

gaaclarke replied:
done

jmagman commented on tools/clang_tidy/lib/clang_tidy.dart:
```diff
@@ -171,23 +189,82 @@ class ClangTidy {
 
+  Iterable<T> _takeShard<T>(Iterable<T> values, int id, int shardCount) sync*{
+    int count = 0;
+    for(final T val in values) {
```
```suggestion
    for (final T val in values) {
```

gaaclarke replied:
done

jmagman commented on tools/clang_tidy/lib/clang_tidy.dart:
```diff
@@ -171,23 +189,82 @@ class ClangTidy {
     return repo.changedFiles;
   }
 
+  Iterable<T> _takeShard<T>(Iterable<T> values, int id, int shardCount) sync*{
```
I guess dart format isn't running on this package?
```suggestion
  Iterable<T> _takeShard<T>(Iterable<T> values, int id, int shardCount) sync* {
```

zanderso replied:
As with the framework repo, dartfmt isn't run in the engine repo.

jmagman commented on tools/clang_tidy/lib/clang_tidy.dart:
```diff
@@ -171,23 +189,82 @@ class ClangTidy {
+    bool allSetsContain(Command command) {
+      for (final List<Command> set in sets) {
+        final Iterable<String> filePaths = set.map((Command e) => e.filePath);
+        if (!filePaths.contains(command.filePath)) {
```
Would the set math be easier making the parameter literally a set `Iterable<Set<Command>>`?

gaaclarke replied:
Like can we get rid of some of this code by taking advantage of Set?  I don't think so.  We'd have to add hashing ability to Command and make clear what it means to look up a Command by file path.  So in that sense it will make the code more complicated.  It might make things a bit faster however.

gaaclarke replied:
Done, moved to `Set<String>` to speed up intersection calculation.

jmagman commented on tools/clang_tidy/lib/clang_tidy.dart:
```diff
@@ -171,23 +189,82 @@ class ClangTidy {
+              .map((dynamic data) =>
+                  Command.fromMap(data as Map<String, dynamic>))
+              .toList());
+      final Iterable<_SetStatusCommand> intersectionResults =
```
This method could use some comments.

gaaclarke replied:
done

jmagman commented on tools/clang_tidy/lib/clang_tidy.dart:
```diff
@@ -171,23 +189,82 @@ class ClangTidy {
+      intersection
+          .sort((Command x, Command y) => x.filePath.compareTo(y.filePath));
+      totalCommands.addAll(
+          _takeShard(intersection, shardId!, 1 + shardBuildCommands.length));
```
Should `shardId` param be nullable if it's going to get `!` here?

jmagman reviewed (commented):
How exactly will `shard-variants` work?  For example will `Mac iOS` be shard 0 and have to build the host as well in order to get the difference between the files?  And the Mac host will be shard 1 and also build iOS?  And then `Linux Android` is shard 0 and builds the Linux host and `Linux host` is shard 1 and also builds Android?  And build (headers), or just gn?

gaaclarke commented:
> How exactly will `shard-variants` work? For example will `Mac iOS` be shard 0 and have to build the host as well in order to get the difference between the files? And the Mac host will be shard 1 and also build iOS? And then `Linux Android` is shard 0 and builds the Linux host and `Linux host` is shard 1 and also builds Android? And build (headers), or just gn?

In order to shard with another variant you have to have its compile_commands.json.  That's generated by running `gn`. 
 In other words, you don't actually have to build the target, just create the build files.

We have 2 options:
1) Shard between [iOS, mac] and [Android, linux]
1) Shard between [iOS, mac, Android, linux]

The second option would be the best, but I'm not sure that it is possible to make gn generate the build files for the platform you are not on, so the first option will be easier to implement.

zanderso reviewed (commented):
Do you have a sense for how many minutes this would improve the `Mac Host clang-tidy` build?

https://ci.chromium.org/p/flutter/builders/prod/Mac%20Host%20clang-tidy

zanderso commented on tools/clang_tidy/lib/clang_tidy.dart:
```diff
@@ -171,23 +189,86 @@ class ClangTidy {
     return repo.changedFiles;
   }
 
+  Iterable<T> _takeShard<T>(Iterable<T> values, int id, int shardCount) sync* {
```
I think the code from here down that is reasoning about sets of commands could be made a bit more idiomatic, and maybe more readable as well, by encapsulating it in a new class (`Commands`, `CommandSet`, or `CommandShard`, or something) that can live in `lib/src/command.dart`.

gaaclarke replied:
Respectfully, I don't think that will help make the code easier to read or maintain.  The small codebase already has poor oop design already and it actually made my job harder.  In objective terms adding more classes will result in more lines of code, more segmentation of the code, and more concepts you need to know to modify the code.

As an example, there is no reason for `CommandSet` to exist when `Set<Command>` (or `List<Command>`) already encapsulates the idea perfectly.  People can come into the codebase and know what that is immediately, not so with `CommandSet`.  I also can't understand from what the difference between a `CommandSet` and a `CommandShard` is.

zanderso replied:
> As an example, there is no reason for `CommandSet` to exist when `Set<Command>` (or `List<Command>`) already encapsulates the idea perfectly.

The code in `getLintCommandsForFiles` is not operating on the `Set`, `List`, etc. in its full generality. Instead it's doing something very specific: trying to filter unwanted items out of a big set. The specific internal data structures that are used to accomplish that goal are an implementation detail. This code would be easier to read and maintain if the the implementation details were encapsulated somehow. It doesn't necessarily have to be a class. It could be more helper methods, for example. I don't have a strong opinion there. However, I think even with comments, the code in `getLintCommandsForFiles` could be improved on, and that is overall what I'm asking for.

> I also can't understand from what the difference between a `CommandSet` and a `CommandShard` is.

These were just naming suggestions for a single new class.

gaaclarke replied:
Yea, I think the documentation could be better and will help with that.  I've updated the docstring in `getLintCommandsForFiles` and added one for `_takeShard`.

zanderso commented on tools/clang_tidy/lib/clang_tidy.dart:
```diff
@@ -171,23 +189,86 @@ class ClangTidy {
+    } else {
+      totalCommands.addAll(buildCommandsData.map((Object? data) => Command.fromMap((data as Map<String, Object?>?)!)));
+    }
+    Stream<Command> filterCommands() async* {
```
Please don't add new uses of `Stream`s / `async*`.

gaaclarke replied:
Are you referring to the guidelines for editing code in the framework?  That doesn't apply here, right?

zanderso replied:
Dart conventions from the framework repo do apply here, yes. But independent of that, `Streams` and `async*` make code very difficult to debug, so should be avoided in general.

gaaclarke replied:
Done.  I think the framework repo recommendation lists alternatives to streams that don't exist here.  Nevertheless, I removed the `Stream` and the `async*`.

gaaclarke commented:
> Do you have a sense for how many minutes this would improve the `Mac Host clang-tidy` build?
> 
> https://ci.chromium.org/p/flutter/builders/prod/Mac%20Host%20clang-tidy

I didn't count the union, just eyeballed it.  Napkin math: maybe there are 100 files overlapped.  I measured clang-tidy to take 16s the other day on my computer.  So for sharding between 2 computers: (100/2 * 16s * (1m / 60s)) = 13 minutes per run.

My computer is faster and the virtual machine is probably under more load than mine so your milage may vary.

zanderso commented:
> > Do you have a sense for how many minutes this would improve the `Mac Host clang-tidy` build?
> > https://ci.chromium.org/p/flutter/builders/prod/Mac%20Host%20clang-tidy
> 
> I didn't count the union, just eyeballed it. Napkin math: maybe there are 100 files overlapped. I measured clang-tidy to take 16s the other day on my computer. So for sharding between 2 computers: (100/2 * 16s * (1m / 60s)) = 13 minutes per run.
> 
> My computer is faster and the virtual machine is probably under more load than mine and technically the tasks are parallelized so your milage may vary.
> 
> Edit: also subtract the extra time needed to run GN an extra time which isn't that long, a couple of seconds? a minute max?

Have you done a local run with the sharding flags? You could also hardcode some values in this PR to try them on the presubmit check.

gaaclarke commented:
> Have you done a local run with the sharding flags? You could also hardcode some values in this PR to try them on the presubmit check.

I didn't, that would take over an hour to run locally and wouldn't give us good results compared to the bots.  I can try to hack the bots to do it in CI.  It's a bit tricky since I'll need to add an extra GN step depending on what platform is executing.  Sounds like you want those numbers before approving though?

gaaclarke commented:
Okay, I've pushed up a hack that may work for the mac host tidy.  The previous run was 65minutes, now I'm running it as shard 0 with the ios_debug shard-variant.

gaaclarke commented:
Heres the overhead from the extra GN call: 5s
```
Generating GN files in: out/ios_debug
Generating Xcode projects took 322ms
Generating compile_commands took 76ms
Done. Made 609 targets from 260 files in 4985ms
```

gaaclarke commented:
Calculating the intersection took longer than expected.  It didn't get measured but it might be worth using a Set think jenn thought.

gaaclarke commented:
Sharding removed 226 files from the mac host clang-tidy job. (836->610)

gaaclarke commented:
@zanderso this run was 43m, so from 65m that's a 22m reduction.  Hey my napkin math wasn't that far off.

gaaclarke commented:
Remember too from my post mortem that the sharding isn't just about saving time either, it's about being in a position where clang-tidy fix patches don't overlap as well.

gaaclarke commented:
Actually, looks like those run time include time queued up, if you look at the actual run times it's `47 mins 44 secs` versus `42 mins 18 secs`.  So more like 5.5 minutes.  We'll see what number we get now that we are using the Sets.  I would only expect a couple minutes difference.

gaaclarke commented:
(with the set's in place it still took ~7m to calculate the intersection)

gaaclarke commented:
Actually ran 1m slower with the Set change.  I have to believe that's a fluke.  Looks like the largest overhead from sharding is parsing the json files.  Anyways, 5.5m savings, final answer.

zanderso commented on tools/clang_tidy/lib/clang_tidy.dart:
```diff
@@ -171,23 +189,91 @@ class ClangTidy {
+  /// This returns a `_SetStatusCommand` for each [Command] in [items].
+  /// `Intersection` if the Command shows up in [items] and its filePath in all
+  /// [filePathSets], otherwise `Difference`.
+  Iterable<_SetStatusCommand> _calcIntersection(
```
The following is more idiomatic. You can use `Iterable.every` in `allSetsContain`, and you can use a "collection-for" instead of a `sync*`:
```dart
Iterable<_SetStatusCommand> _calcIntersection(
  Iterable<Commands> items,
  Iterable<Set> filePathSets,
) {
  bool allSetsContain(Command command) {
    return filePathSets.every(
      (Set<String> filePathSet) => filePathSet.contains(command.filePath),
    );
  }
  return <_SetStatusCommand>[
    for (final Command command in items)
      if (allSetsContain(command))
        _SetStatusCommand(_SetStatus.Intersection, command)
      else
        _SetStatusCommand(_SetStatus.Difference, command),
  ];
}
```

zanderso commented on tools/clang_tidy/lib/clang_tidy.dart:
```diff
@@ -109,12 +120,19 @@ class ClangTidy {
-    ) as List<dynamic>;
-    final List<Command> changedFileBuildCommands = await getLintCommandsForChangedFiles(
+    ) as List<Object?>;
+    final List<List<Object?>> shardBuildCommandsData = options
```
"collection-for" is more idiomatic:
```dart
final List<List<Object?>> shardBuildCommandsData = <List<Object?>>[
  for (io.File file in options.shardCommandsPaths)
    jsonDecode(file.readAsStringSync()) as List<Object?>,
];
```

zanderso commented on tools/clang_tidy/lib/clang_tidy.dart:
```diff
@@ -171,23 +189,91 @@ class ClangTidy {
+    int? shardId,
+  ) {
+    final List<Command> totalCommands = <Command>[];
+    if (sharedBuildCommandsData.isNotEmpty) {
```
Re-writing with collection-for I think is an improvement.

```dart
if (sharedBuildCommandsData.isNotEmpty) {
  final Iterable<Command> buildCommands = <Command>[
    for (Object? data in buildCommandsData)
      Command.fromMap(data as Map<String, Object?>),
  ];
  final Iterable<Set<String>> shardFilePaths = <Set<String>>[
    for (List<Object?> list in  sharedBuildCommandsData) {
      for (Object? data in list)
        Command.fromMap((data as Map<String, Object?>?)!).filePath,
    },
  ];
  final Iterable<_SetStatusCommand> intersectionResults = _calcIntersection(
    buildCommands, shardFilePaths,
  );
  totalCommands.addAll(<Command>[
    for (_SetStatusCommand element in intersectionResults)
      if (element.setStatus == _SetStatus.Difference)
        element.command,
  ]);
  final List<Command> intersection = <Command>[
    for (_SetStatusCommand element in intersectionResults)
      if (element.setStatus == _SetStatus.Intersection)
        element.command,
  ];
  totalCommands.addAll(_takeShard(
    intersection,
    shardId!,
    1 + sharedBuildCommandsData.length,
  ));
}
```

zanderso commented on tools/clang_tidy/lib/src/options.dart:
```diff
@@ -87,25 +93,42 @@ class Options {
-    );
+    buildCommandsPath ??= variantToBuildCommandsFilePath(argResults['target-variant'] as String);
     final io.File buildCommands = io.File(buildCommandsPath);
+    final List<io.File> shardCommands =
```
```dart
final String shardVariants = argResults['shard-variants'] as String? ?? '';
final List<io.File> shardCommands = <io.File>[
  for (String element in buildCommands.split(','))
    if (element.isNotEmpty)
      io.FIle(variantToBuildCommandsFilePath(element)),
];
```

gaaclarke commented on tools/clang_tidy/lib/clang_tidy.dart:
```diff
@@ -171,23 +189,91 @@ class ClangTidy {
+              element.setStatus == _SetStatus.Intersection)
+          .map((_SetStatusCommand e) => e.command)
+          .toList();
+      // Make sure to sort results, not sure if there is a defined order in the json file.
```
It matters because it's how the sharding scheme works.  We have to guarantee that all shards are starting with the same order to make sure that their jobs don't overlap.

zanderso replied:
Okay, I think I'm missing something big. Could you explain a bit more about why the order matters, or what would go wrong if the jobs are out of order?

gaaclarke replied:
Yep, so imagine we are sharding across iOS and mac.  They both generated their compile_commands.json files.  They look like this:
iOS
```
[
  { "path": "foo.cc" },
  { "path": "bar.cc" },
]
```
mac
```
[
  { "path": "bar.cc" },
  { "path": "foo.cc" },
]
```

If you ran mac as shard 0 and ios and shard 1 (where job(n) = n * 2 + shard), they will both just execute on "foo.cc".  We don't have a guarantee about the order of compile_command.json as far as I know, so they have to be sorted.

zanderso replied:
Got it. This explanation should probably go in a comment in the code.

gaaclarke replied:
expanded on the existing comment

gaaclarke commented:
> Sorry to keep spinning on this PR. Main comment is to use 'collection-for' instead of `map`, and `where` to be more idiomatic with modern Dart. I've sketched out what it would look like below, but I haven't actually tried to compile and run this code.

Thanks for expanding on your comments about idiomatic code, now I understand better what you were thinking.  The changes you are suggesting while logically equivalent are not equivalent in performance.  It isn't a matter of code style.  Let me explain.

I've specifically gone out of my way to use Dart iterators here in order to optimize the code to avoid any extra work and to reduce copying and memory usage.  There are instances where the algorithm will short circuit (ex when it finds that a file isn't in the intersection) where using iterators will allow us to avoid all the work that follows.  Also, it allows the code not be be memory bound which is future proof against whatever happens to compile_commands.json in the future.

I'm happy to explain further, but I think you'll understand after I've pointed it out.

zanderso commented:
> I've specifically gone out of my way to use Dart iterators here in order to optimize the code to avoid any extra work and to reduce copying and memory usage.

This smells like premature optimization to me. Unless there's a known performance bug that we're fixing, I think writing idiomatic code should take priority.

zanderso commented:
> It's not, json file to output takes ~7 minutes to execute already. 

Could you clarify what are the operations that are taking the 7 minutes?

gaaclarke commented:
> > It's not, json file to output takes ~7 minutes to execute already.
> 
> Could you clarify what are the operations that are taking the 7 minutes?

I haven't investigated it beyond the fact that it happens between the clang tidy banner getting printed out and the list of files that are being linted is printed out.  So the combination of parsing 2 12MB files and this algorithm I added doing set math on 2 sets that are about 1k big is probably it.  My money is on the JSON parsing.  I haven't had a chance to investigate, I've been having conversations about what is idiomatic dart code lol ;)

zanderso commented:
> I haven't investigated it beyond the fact that it happens between the clang tidy banner getting printed out and the list of files that are being linted is printed out.

Got it. Sorry for the distraction. It sounds like if we could dig into that a bit, and debug it, then this change could be an even bigger win.

gaaclarke commented:
I was wrong, locally parsing the json files takes 0.13s, `getLintCommandsForFiles` takes 3m01s.  I wasn't planning on investigating this unless I had time since the PR still represents a 5.5m savings.  I can give this a quick look to see if I can speed up since we already started cracking it open and is apropos the comments on the code.

gaaclarke commented:
I've removed instances of calling `.map`  with dart's collection-for.  I don't agree the code is easier to maintain or map isn't idiomatic but it was found that there is a performance bug in Dart so I switched them over.  I'll file a bug against Dart for it as well.  I've also turned the lint.sh hack on so we can measure the time it takes to run the step now.

gaaclarke commented:
Okay, with the new change the task is 38m and counting.  I verified that it is only linting 610 files, so the sharding is working.  This points the the flaw in doing performance benchmarks with sample size 1 on a machine we can't control.  Who knows what's going on with the machine.

gaaclarke commented:
Turned off the ci hack, merged master.

gaaclarke commented on tools/clang_tidy/test/clang_tidy_test.dart:
```diff
@@ -205,10 +205,77 @@ Future<int> main(List<String> args) async {
+          'command': '../../buildtools/mac-x64/clang/bin/clang $filePath',
+          'file': filePath,
+        };
+    final List<String> filePaths = () sync* {
```
done

gaaclarke commented:
Here is the CL for enabling the sharding in the recipe: https://flutter-review.googlesource.com/c/recipes/+/35760

gaaclarke commented:
I just did the math on the final improvement to Mac Host clang-tidy, averaging the time of 5 samples before and after this was turned on.

The average time went from 37.6m to 29.8m, a **reduction of 7.8 minutes each run**, or a 20% reduction in time.

before samples:
- https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20Host%20clang-tidy/658/overview
- https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20Host%20clang-tidy/659/overview
- https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20Host%20clang-tidy/660/overview
- https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20Host%20clang-tidy/661/overview
- https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20Host%20clang-tidy/662/overview

after samples:
- https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20Host%20clang-tidy/679/overview
- https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20Host%20clang-tidy/678/overview
- https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20Host%20clang-tidy/677/overview
- https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20Host%20clang-tidy/676/overview
- https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20Host%20clang-tidy/675/overview
